src/main.rs

```diff
@@ -213,37 +234,36 @@ fn reader_for(path: &Path) -> anyhow::Result<BufReader<File>> {
 fn load_cert_and_key(
 	cert_path: &Path,
 	key_path: &Path
-) -> anyhow::Result<(Vec<Certificate>, PrivateKey)> {
-	let certificates = rustls_pemfile::read_all(&mut reader_for(cert_path)?)
-		.with_context(|| format!("Failed to parse {}", cert_path.display()))?
-		.iter()
-		.filter_map(|cert| match cert {
-			rustls_pemfile::Item::X509Certificate(cert) => {
-				Some(Certificate(cert.to_owned()))
-			},
-			_ => None
+) -> anyhow::Result<Arc<impl ResolvesServerCert>> {
+	let certificates = rustls_pemfile::certs(&mut reader_for(cert_path)?)
+		.filter_map(|cert| {
+			match cert.with_context(|| format!("Failed to parse {}", cert_path.display()))
+			{
+				Ok(s) => Some(s),
+				Err(_) => None
+			}
 		})
 		.collect::<Vec<_>>();
 	if certificates.is_empty() {
 		bail!("No x509 certificate found in {}", cert_path.display());
 	}
-	let key = rustls_pemfile::read_all(&mut reader_for(key_path)?)
+	let key = rustls_pemfile::private_key(&mut reader_for(key_path)?)
 		.with_context(|| format!("Failed to parse {}", key_path.display()))?
-		.iter()
-		.find_map(|item| match item {
-			rustls_pemfile::Item::ECKey(key)
-			| rustls_pemfile::Item::RSAKey(key)
-			| rustls_pemfile::Item::PKCS8Key(key) => Some(key),
-			_ => None
-		})
-		.map(|key| PrivateKey(key.to_owned()))
 		.ok_or_else(|| {
 			anyhow!(
 				"No private RSA/PKCS8/ECKey key found in {}",
 				key_path.display()
 			)
 		})?;
-	Ok((certificates, key))
+
+	let certified_key = CertifiedKey::from_der(
+		certificates,
+		key,
+		CryptoProvider::get_default().ok_or_else(|| {
+			anyhow!("CryptoProvider default should have been registered in main!")
+		})?
```

Comment:
I think this could just as well be a `.expect()`. There is no way to recover a missing crypto provider.